### PR TITLE
core, inprocess, util: move inprocess and util code into their own new artifacts grpc-inprocess and grpc-util

### PR DESCRIPTION
**Pull request by sanjaypujare**
CC @ejona86

### CODE REVIEW

**Review by larry-safran — APPROVED** [review]

**Comment by sanjaypujare** [comment]
Thanks @larry-safran . Now waiting for @ejona86 's LGTM....

**Comment by sanjaypujare** [comment]
Gentle ping @ejona86

File: core/BUILD.bazel [hunk 1]
```diff
@@ -4,33 +4,16 @@ java_library(
...
-        "@com_google_code_findbugs_jsr305//jar",
-        "@com_google_errorprone_error_prone_annotations//jar",
-        "@com_google_guava_guava//jar",
-        "@com_google_j2objc_j2objc_annotations//jar",
-    ],
-)
-
 java_library(
     name = "internal",
+    visibility = ["//visibility:public"],
```
Comment (ejona86) [inline core/BUILD.bazel:9]: internal should not be public. Why did you need this? It was already allowing subpackages.

File: core/src/main/resources/META-INF/services/io.grpc.LoadBalancerProvider [hunk 2]
```diff
@@ -1,3 +1 @@
 io.grpc.internal.PickFirstLoadBalancerProvider
-io.grpc.util.SecretRoundRobinLoadBalancerProvider$Provider
```
Comment (ejona86) [inline core/src/main/resources/META-INF/services/io.grpc.LoadBalancerProvider:2]: RR moving to grpc-util could definitely break users in a hard-to-diagnose way. I'm not sure how we want to handle this. Short-term we could have a runtimeDependency in grpc-core onto grpc-util.

That would mean we'd also want `//inprocess` in `core_maven`

File: core/BUILD.bazel [hunk 3]
```diff
@@ -4,33 +4,16 @@ java_library(
     exports = ["//api"],
 )
 
-java_library(
-    name = "inprocess",
```
Comment (ejona86) [inline core/BUILD.bazel:8]: Leave a forwarding library in place. Keep `name` and `visibility` and have it export //inprocess. Ditto for util.

File: inprocess/build.gradle [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+        sourceSets.main,
+        sourceSets.test
+    ]
+}
+
+components.java.withVariantsFromConfiguration(configurations.testFixturesApiElements) { skip() }
+components.java.withVariantsFromConfiguration(configurations.testFixturesRuntimeElements) { skip() }
+
+
+tasks.register("versionFile") {
```
Comment (ejona86) [inline inprocess/build.gradle:61]: Keep this in core. It is not needed in inprocess and util.

File: inprocess/build.gradle [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+
+    signature libraries.signature.java
+    signature libraries.signature.android
+}
+
+tasks.named("javadoc").configure {
+    exclude 'io/grpc/inprocess/Internal*'
+}
+
+animalsniffer {
```
Comment (ejona86) [inline inprocess/build.gradle:49]: Delete. There's no jmh.

File: inprocess/build.gradle [hunk 6]
```diff
@@ -0,0 +1,65 @@
+plugins {
+    id "java-library"
+    id "java-test-fixtures"
```
Comment (ejona86) [inline inprocess/build.gradle:3]: Delete. There are no test fixtures here.

File: inprocess/build.gradle [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+
+animalsniffer {
+    // Don't check sourceSets.jmh
+    sourceSets = [
+        sourceSets.main,
+        sourceSets.test
+    ]
+}
+
+components.java.withVariantsFromConfiguration(configurations.testFixturesApiElements) { skip() }
```
Comment (ejona86) [inline inprocess/build.gradle:57]: Delete both lines

File: inprocess/build.gradle [hunk 8]
```diff
@@ -0,0 +1,65 @@
+plugins {
+    id "java-library"
+    id "java-test-fixtures"
+    id "maven-publish"
+
+    id "me.champeau.jmh"
```
Comment (ejona86) [inline inprocess/build.gradle:6]: Delete.

File: inprocess/build.gradle [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+    id "ru.vyarus.animalsniffer"
+}
+
+import static java.nio.charset.StandardCharsets.US_ASCII;
```
Comment (ejona86) [inline inprocess/build.gradle:10]: Delete both imports.

File: inprocess/build.gradle [hunk 10]
```diff
@@ -0,0 +1,65 @@
...
+
+import com.google.common.primitives.Bytes;
+
+description = 'gRPC: Inprocess'
+
+dependencies {
+    api project(':grpc-core')
+    // force dependent jars to depend on latest grpc-context
```
Comment (ejona86) [inline inprocess/build.gradle:18]: Remove this. It was only for core.

File: inprocess/build.gradle [hunk 11]
```diff
@@ -0,0 +1,65 @@
...
+    runtimeOnly project(":grpc-context")
+    implementation libraries.gson,
```
Comment (ejona86) [inline inprocess/build.gradle:20]: Delete all these deps except guava.

File: inprocess/build.gradle [hunk 12]
```diff
@@ -0,0 +1,65 @@
...
+            libraries.guava,
+            libraries.perfmark.api
+    testFixturesApi libraries.junit
+    testFixturesImplementation libraries.guava,
+            libraries.mockito.core,
+            libraries.truth,
+            project(':grpc-testing')
+    testImplementation testFixtures(project(':grpc-api')),
+            project(':grpc-testing'),
+            project(':grpc-grpclb'),
```
Comment (ejona86) [inline inprocess/build.gradle:33]: Delete.

File: inprocess/build.gradle [hunk 13]
```diff
@@ -0,0 +1,65 @@
...
+            libraries.animalsniffer.annotations,
+            libraries.errorprone.annotations,
+            libraries.guava,
+            libraries.perfmark.api
+    testFixturesApi libraries.junit
+    testFixturesImplementation libraries.guava,
+            libraries.mockito.core,
+            libraries.truth,
+            project(':grpc-testing')
+    testImplementation testFixtures(project(':grpc-api')),
```
Comment (ejona86) [inline inprocess/build.gradle:31]: Delete grpc-api fixture

File: inprocess/build.gradle [hunk 14]
```diff
@@ -0,0 +1,65 @@
...
+            project(':grpc-testing'),
+            project(':grpc-grpclb'),
+            testFixtures(project(':grpc-core'))
+    testImplementation libraries.guava.testlib
+
+    testRuntimeOnly project(':grpc-census')
```
Comment (ejona86) [inline inprocess/build.gradle:37]: Delete.

File: repositories.bzl [hunk 15]
```diff
@@ -79,6 +79,8 @@ IO_GRPC_GRPC_JAVA_OVERRIDE_TARGETS = {
     "io.grpc:grpc-stub": "@io_grpc_grpc_java//stub",
     "io.grpc:grpc-testing": "@io_grpc_grpc_java//testing",
     "io.grpc:grpc-xds": "@io_grpc_grpc_java//xds:xds_maven",
+    "io.grpc:grpc-inprocess": "@io_grpc_grpc_java//inprocess",
```
Comment (ejona86) [inline repositories.bzl:82]: Sort these.

File: rls/build.gradle [hunk 16]
```diff
@@ -9,8 +9,8 @@ plugins {
 description = "gRPC: RouteLookupService Loadbalancing plugin"
 
 dependencies {
-    implementation project(':grpc-core'),
-            project(':grpc-protobuf'),
+    api project(':grpc-util')
```
Comment (ejona86) [inline rls/build.gradle:12]: This should be an implementation dependency.

File: util/BUILD.bazel [hunk 17]
```diff
@@ -0,0 +1,30 @@
...
+        "@com_google_j2objc_j2objc_annotations//jar",
+        "@org_codehaus_mojo_animal_sniffer_annotations//jar",
+    ],
+)
+
+# Mirrors the dependencies included in the artifact on Maven Central for usage
+# with maven_install's override_targets. Should only be used as a dep for
+# pre-compiled binaries on Maven Central.
+java_library(
+    name = "util_maven",
```
Comment (ejona86) [inline util/BUILD.bazel:24]: There's no need to export //api from here, so delete the entire target. If it were needed, it should have been used from repositories.bzl

File: util/src/bazel-util/resources/META-INF/services/io.grpc.LoadBalancerProvider
Comment (ejona86) [inline util/src/bazel-util/resources/META-INF/services/io.grpc.LoadBalancerProvider:1]: Delete this file and use src/main/resources instead.

File: core/src/main/resources/META-INF/services/io.grpc.LoadBalancerProvider (hunk 2 above)
Comment (sanjaypujare) [reply core/src/main/resources/META-INF/services/io.grpc.LoadBalancerProvider:2]: > ... Short-term we could have a runtimeDependency in grpc-core onto grpc-util.

Interesting. grpc-util has an api dependency on grpc-core so doing what you suggest makes the dependency circular (even with runtimeOnly dependency). Any ideas?

Comment (sanjaypujare) [reply core/src/main/resources/META-INF/services/io.grpc.LoadBalancerProvider:2]: Looks like this is a blocker and needs to be resolved. So I'll wait for that before I address the remaining comments.

Comment (ejona86) [reply core/src/main/resources/META-INF/services/io.grpc.LoadBalancerProvider:2]: I was aware that would be the case. But it should work. Are you saying it doesn't work?

Comment (sanjaypujare) [reply core/src/main/resources/META-INF/services/io.grpc.LoadBalancerProvider:2]: Yes, I had tried it before I added the comment:

```
* What went wrong:
Circular dependency between the following tasks:
:grpc-core:checkUpperBoundDeps
+--- :grpc-core:jar
|    +--- :grpc-core:classes
|    |    \--- :grpc-core:compileJava
|    |         \--- :grpc-core:checkUpperBoundDeps (*)
|    \--- :grpc-core:compileJava (*)
\--- :grpc-util:jar
     +--- :grpc-util:classes
     |    \--- :grpc-util:compileJava
     |         +--- :grpc-core:compileJava (*)
     |         \--- :grpc-util:checkUpperBoundDeps
     |              +--- :grpc-core:jar (*)
     |              \--- :grpc-util:jar (*)
     \--- :grpc-util:compileJava (*)

```

Comment (sanjaypujare) [reply core/src/main/resources/META-INF/services/io.grpc.LoadBalancerProvider:2]: This has been resolved

File: core/BUILD.bazel (hunk 1 above)
Comment (sanjaypujare) [reply core/BUILD.bazel:9]: done

File: core/BUILD.bazel (hunk 3 above)
Comment (sanjaypujare) [reply core/BUILD.bazel:8]: done

**Review by ejona86 — COMMENTED** [review]
Most of my earlier comments apply to `util/build.gradle` as well.

File: inprocess/build.gradle (hunk 8 above)
Comment (sanjaypujare) [reply inprocess/build.gradle]: deleted but not deleting for util (since util has jmh)

File: util/src/bazel-util/resources/META-INF/services/io.grpc.LoadBalancerProvider
Comment (sanjaypujare) [reply util/src/bazel-util/resources/META-INF/services/io.grpc.LoadBalancerProvider:1]: done

File: grpclb/build.gradle [hunk 18]
```diff
@@ -10,7 +10,7 @@ plugins {
 description = "gRPC: GRPCLB LoadBalancer plugin"
 
 dependencies {
-    implementation project(':grpc-core'),
+    implementation project(':grpc-inprocess'),
```
Comment (ejona86) [inline grpclb/build.gradle:13]: Why was it changed to depend on inprocess?

File: util/build.gradle [hunk 19]
```diff
@@ -0,0 +1,34 @@
...
+    id "me.champeau.jmh"
+    id "ru.vyarus.animalsniffer"
+}
+
+description = 'gRPC: Util'
+
+dependencies {
+    api project(':grpc-core')
+
+    implementation libraries.gson,
```
Comment (ejona86) [inline util/build.gradle:14]: Why depend on gson?

File: util/build.gradle [hunk 20]
```diff
@@ -0,0 +1,34 @@
...
+            libraries.animalsniffer.annotations,
```
Comment (ejona86) [inline util/build.gradle:15]: Why depend on animalsniffer?

File: core/build.gradle [hunk 21]
```diff
@@ -23,7 +23,8 @@ description = 'gRPC: Core'
 dependencies {
     api project(':grpc-api')
     // force dependent jars to depend on latest grpc-context
-    runtimeOnly project(":grpc-context")
+    runtimeOnly project(":grpc-context"),
+            project(":grpc-util")
```
Comment (ejona86) [inline core/build.gradle:27]: Include a comment that we depend on grpc-util to pull in RR

File: core/build.gradle [hunk 22]
```diff
@@ -23,7 +23,8 @@ description = 'gRPC: Core'
...
     implementation libraries.gson,
             libraries.android.annotations,
             libraries.animalsniffer.annotations,
```
Comment (ejona86) [inline core/build.gradle:30]: Delete

File: util/build.gradle (hunk 20 above)
Comment (sanjaypujare) [reply util/build.gradle:15]: Without it I get

```
grpc-java/util/src/main/java/io/grpc/util/AdvancedTlsX509TrustManager.java:41: error: package org.codehaus.mojo.animal_sniffer does not exist
import org.codehaus.mojo.animal_sniffer.IgnoreJRERequirement;
                                       ^
grpc-java/util/src/main/java/io/grpc/util/AdvancedTlsX509TrustManager.java:49: error: cannot find symbol
@IgnoreJRERequirement
 ^
  symbol: class IgnoreJRERequirement
2 errors

```

File: core/build.gradle (hunk 22 above)
Comment (sanjaypujare) [reply core/build.gradle:30]: Without it I get:

```
core/src/main/java/io/grpc/internal/JndiResourceResolverFactory.java:38: error: package org.codehaus.mojo.animal_sniffer does not exist
import org.codehaus.mojo.animal_sniffer.IgnoreJRERequirement;
                                       ^
ore/src/main/java/io/grpc/internal/JndiResourceResolverFactory.java:197: error: cannot find symbol
  @IgnoreJRERequirement
   ^
  symbol:   class IgnoreJRERequirement
  location: class JndiResourceResolverFactory
2 errors

```

File: util/build.gradle (hunk 19 above)
Comment (sanjaypujare) [reply util/build.gradle:14]: Without it the build of grpc-services fails:

```
* What went wrong:
Execution failed for task ':grpc-services:checkUpperBoundDeps'.
> Maven version skew: com.google.code.gson:gson (2.8.9 != 2.10.1) Bad version dependency path: [project ':grpc-services', com.google.protobuf:protobuf-java-util:3.22.3] Run './gradlew :grpc-services:dependencies --configuration runtimeClasspath' to diagnose
```

Comment (sanjaypujare) [reply util/build.gradle:14]: And then running `./gradlew :grpc-services:dependencies --configuration runtimeClasspath` (as suggested) gives me:

```
runtimeClasspath - Runtime classpath of source set 'main'.
...
+--- project :grpc-util
|    +--- project :grpc-core
|    |    +--- project :grpc-api (*)
|    |    +--- com.google.code.gson:gson:2.10.1 ===> v2.10.1 referenced
|    |    +--- com.google.android:annotations:4.1.1.4
|    |    +--- org.codehaus.mojo:animal-sniffer-annotations:1.23
|    |    +--- com.google.errorprone:error_prone_annotations:2.18.0
|    |    +--- com.google.guava:guava:32.0.1-android -> 32.0.1-jre (*)
|    |    +--- io.perfmark:perfmark-api:0.26.0
|    |    +--- project :grpc-context
|    |    |    \--- project :grpc-api (*)
|    |    \--- project :grpc-util (*)
|    +--- org.codehaus.mojo:animal-sniffer-annotations:1.23
|    \--- com.google.guava:guava:32.0.1-android -> 32.0.1-jre (*)
+--- com.google.protobuf:protobuf-java-util:3.22.3
|    +--- com.google.protobuf:protobuf-java:3.22.3
|    +--- com.google.code.findbugs:jsr305:3.0.2
|    +--- com.google.code.gson:gson:2.8.9 -> 2.10.1 ===> v2.8.9 referenced 
|    +--- com.google.errorprone:error_prone_annotations:2.11.0 -> 2.18.0
|    +--- com.google.guava:guava:31.1-jre -> 32.0.1-jre (*)
|    \--- com.google.j2objc:j2objc-annotations:1.3 -> 2.8
...
```

I can also fix it by making `libraries.gson` a runtimeOnly dependency of `grpc-util` and then add a comment describing why it is required?

Comment (ejona86) [reply util/build.gradle]: > I can also fix it by making libraries.gson a runtimeOnly dependency of grpc-util and then add a comment describing why it is required?

Yes. Do that for forced upgrading to fix version skew.

File: util/build.gradle (hunk 20 above)
Comment (ejona86) [reply util/build.gradle]: Looks like it is appropriate, because we are using it. Maybe another dependency had pulled it in when I had tried, but we've now cleaned up that dependency.

File: util/build.gradle (hunk 19 above)
Comment (sanjaypujare) [reply util/build.gradle]: Already done with a comment

File: core/BUILD.bazel [hunk 23]
```diff
@@ -71,9 +47,8 @@ java_library(
     name = "core_maven",
     visibility = ["//visibility:public"],
     exports = [
-        ":inprocess",
         ":internal",
-        ":util",
         "//api",
+        "//inprocess",
```
Comment (ejona86) [inline core/BUILD.bazel:52]: This should have exported //util, not //inprocess (to mirror core/build.gradle)

File: util/build.gradle [hunk 24]
```diff
@@ -0,0 +1,34 @@
...
+}
+
+description = 'gRPC: Util'
+
+dependencies {
+    api project(':grpc-core')
+
+    implementation libraries.animalsniffer.annotations,
+            libraries.guava
+    runtimeOnly libraries.gson  // to fix checkUpperBoundDeps error in services
```
Comment (ejona86) [inline util/build.gradle:16]: No, this should go in services.

**Review by ejona86 — COMMENTED** [review]
LGTM

File: core/src/main/resources/META-INF/services/io.grpc.LoadBalancerProvider (hunk 2 above)
Comment (mdering) [reply core/src/main/resources/META-INF/services/io.grpc.LoadBalancerProvider:2]: hi i just wanted to let you know that this did indeed break for us in a hard to diagnose way. Is there a correct way to upgrade this library so that it does not break? in general we use something like
```java
ManagedChannelBuilder
            .forTarget(URL)
            .defaultLoadBalancingPolicy("round_robin")
            .usePlaintext()
            .build()
```

Comment (ejona86) [reply core/src/main/resources/META-INF/services/io.grpc.LoadBalancerProvider:2]: @mdering "break" is not a helpful description. Please open a new issue and describe the error/problem you see when upgrading. You can link to this PR. For this specific change, it's also important whether you are shading/shadowing/making a fat jar.
